### PR TITLE
Add rule function for IP addresses, only supports IPv4 for now

### PR DESCRIPTION
**Pull request by kemasdimas**
Referring to #45

### CODE REVIEW

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 1]
```diff
@@ -274,6 +274,63 @@ private sealed class RuleExpr<out T> {
         }
         //endregion
 
+        //region IP matching\
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]: Regions currently map to “Type” from the function table. We should move `endregion` below it.

(it should be “Lookup” above this, but it's unrelated with this PR)

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 2]
```diff
@@ -274,6 +274,63 @@ private sealed class RuleExpr<out T> {
...
+
+        data class MatchesIpv4(
+            val range: RuleExpr<String>,
+            val value: RuleExpr<String>
+        ) : FunctionExpr<Boolean>() {
+            override fun eval(env: JsonObject): Boolean {
+                val ipRange = ipv4Regex.matchEntire(range.castEval(env))
+                val ipValue = ipv4Regex.matchEntire(value.castEval(env))
+
+                if (ipRange == null || ipRange.groups[1] == null)
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]: This and the line below are missing braces. Detekt is [set up automatically](https://github.com/Doist/ffs/blob/ebb828477e93335f7488d64b7105555742fedef7/build.gradle.kts#L128-L134), but it might not run from GUIs.

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 3]
```diff
@@ -281,6 +281,27 @@ class RuleEvalTest {
...
     }
 
+    @Test
+    fun ipv4matches() {
+        assertEquals(0f, eval("""ipv4matches("10.0.0.0/28", "10.0.0.16")"""))
+        assertEquals(0f, eval("""ipv4matches("10.0.0.0/28", "9.255.255.255")"""))
+        assertEquals(0f, eval("""ipv4matches("192.168.2.10/24", "192.168.3.255")"""))
+        assertEquals(1f, eval("""ipv4matches("10.0.0.0/28", "10.0.0.15")"""))
+        assertEquals(1f, eval("""ipv4matches("255.0.0.0/28", "255.0.0.4")"""))
+        assertEquals(1f, eval("""ipv4matches("192.168.2.10/24", "192.168.2.254")"""))
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]: Since IPs and masks work with ranges, it would be great to test the boundaries of what should work (here) and what shouldn't (below) more thoroughly.

**Review by goncalossilva — CHANGES_REQUESTED** [review]
Thanks, @kemasdimas! This is in the right direction, and it's great that you updated the documentation too. There are 2 considerations I wanted to discuss before we advance.

**Ergonomics**
Could it be ergonomic and flexible to leverage the existing composability of the existing functions, instead of having multiple `match` variants? For instance:
- `contains(cidr("10.0.0.0/28"), ip("10.0.0.16"))`
- `lt(ip(env["user.ip"], ip("109.71.47.255"))`
- `not(ip(env["user.ip"], ip("127.0.0.1"))`

Of course `ip` and `cidr` would have to return something. `Long` and `Range` could be good choices, but this is related to the next point.

**Performance**
The current proposal relies on a non-trivial regular expression that we can't pre-compile, which can be suboptimal considering FFS is designed to do thousands of evaluations per second. But IP addresses are just numbers, and IP+mask are just ranges. We could treat them like that?

Specifically, what if we:
- Add a [range type](https://gist.github.com/goncalossilva/2406eb278eb0c8a1b89c3db76f4e0e2b) (with tests & docs). It seems useful, regardless of IPs.
- Implement `ip(value)` as a function that takes an IP, and returns a `Long`. An IP is just a group of 4 bytes, so we could parse it ([example](https://stackoverflow.com/questions/62124452/converting-cidr-notation-into-ip-range-without-other-library)) and combine it together in a single number.
- Implement `cidr(value)` as a function that takes an IP+mast, and returns a range. It can reuse the range and `ip(value)` implementations from before.

This could be very efficient. What do you think?

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt (hunk 2 above)
Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]: Interesting. Detekt doesn't catch this. Still, we always use braces unless the condition and statement fit on a single line (which we also tend to avoid except for early returns) per [Kotlin's style guide](https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements).

Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]: Thanks for pointing this out 😊

Yes, Detekt didn't catch this one when I tried to push to commit. I will review the coding conventions and update my code formatting properly

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt (hunk 3 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt:291]: > **Ergonomics** Could it be ergonomic and flexible to leverage the existing composability of the existing functions, instead of having multiple `match` variants? For instance:
> 
> * `contains(cidr("10.0.0.0/28"), ip("10.0.0.16"))`
> * `lt(ip(env["user.ip"], ip("109.71.47.255"))`
> * `not(ip(env["user.ip"], ip("127.0.0.1"))`
> 
> Of course `ip` and `cidr` would have to return something. `Long` and `Range` could be good choices, but this is related to the next point.

I agree, I haven't put more thought into this, and creating the rule to be composable seems to be a lot better. So there will be 2 new function

cidr -> return `Range`
ip -> return `Long`

> **Performance** The current proposal relies on a non-trivial regular expression that we can't pre-compile, which can be suboptimal considering FFS is designed to do thousands of evaluations per second. But IP addresses are just numbers, and IP+mask are just ranges. We could treat them like that?

In this case, the regex is redundant, it was only used to validate the IP / CIDR format. Referring to the Ergonomics, treating IP addresses and mask as numbers are to be expected.

> Specifically, what if we:
> 
> * Add a [range type](https://gist.github.com/goncalossilva/2406eb278eb0c8a1b89c3db76f4e0e2b) (with tests & docs). It seems useful, regardless of IPs.
> * Implement `ip(value)` as a function that takes an IP, and returns a `Long`. An IP is just a group of 4 bytes, so we could parse it ([example](https://stackoverflow.com/questions/62124452/converting-cidr-notation-into-ip-range-without-other-library)) and combine it together in a single number.
> * Implement `cidr(value)` as a function that takes an IP+mast, and returns a range. It can reuse the range and `ip(value)` implementations from before.
> 
> This could be very efficient. What do you think?

Yes, this will be efficient to run. My only concern is the IPv6 format. It's not straightforward to convert the IP to number, even using ULong, so we'll need to find a way to represent the IPv6 in 1 single variable while keeping it comparable to the CIDR

For the Alpha milestone, will it be enough if we only support IPv4?

**Comment by kemasdimas** [comment]
> **Ergonomics** Could it be ergonomic and flexible to leverage the existing composability of the existing functions, instead of having multiple `match` variants? For instance:
> 
> * `contains(cidr("10.0.0.0/28"), ip("10.0.0.16"))`
> * `lt(ip(env["user.ip"], ip("109.71.47.255"))`
> * `not(ip(env["user.ip"], ip("127.0.0.1"))`
> 
> Of course `ip` and `cidr` would have to return something. `Long` and `Range` could be good choices, but this is related to the next point.

I agree, I haven't put more thought into this, and creating the rule to be composable seems to be a lot better. So there will be 2 new function

cidr -> return `Range`
ip -> return `Long`

> **Performance** The current proposal relies on a non-trivial regular expression that we can't pre-compile, which can be suboptimal considering FFS is designed to do thousands of evaluations per second. But IP addresses are just numbers, and IP+mask are just ranges. We could treat them like that?

In this case, the regex is redundant, it was only used to validate the IP / CIDR format. Referring to the Ergonomics, treating IP addresses and mask as numbers are to be expected.

> Specifically, what if we:
> 
> * Add a [range type](https://gist.github.com/goncalossilva/2406eb278eb0c8a1b89c3db76f4e0e2b) (with tests & docs). It seems useful, regardless of IPs.
> * Implement `ip(value)` as a function that takes an IP, and returns a `Long`. An IP is just a group of 4 bytes, so we could parse it ([example](https://stackoverflow.com/questions/62124452/converting-cidr-notation-into-ip-range-without-other-library)) and combine it together in a single number.
> * Implement `cidr(value)` as a function that takes an IP+mast, and returns a range. It can reuse the range and `ip(value)` implementations from before.
> 
> This could be very efficient. What do you think?

Yes, this will be efficient to run. My only concern is the IPv6 format. It's not straightforward to convert the IP to number, even using ULong, so we'll need to find a way to represent the IPv6 in 1 single variable while keeping it comparable to the CIDR

For the Alpha milestone, will it be enough if we only support IPv4?

**Comment by goncalossilva** [comment]
128-bit can represent IPv6, so if this strategy works well, we can consider using a List with two Long, or representing IPs using byte arrays, os similar. But one challenge at a time, IPv4 for now sounds good! 😄

**Comment by kemasdimas** [comment]
First thing, thank you for taking your time to review my code, and really appreciate all your detailed writings, I learned a lot of new things and processes from this codebase 😊👍

**Summary**
1. Implement a Range function based on https://gist.github.com/goncalossilva/2406eb278eb0c8a1b89c3db76f4e0e2b, with a small tweak to allow the range to use a function, e.g. `contains([datetime("2021-1-1"):datetime("2022-10-1")], now())`
2. Create IP to decimal converter function that can be used in rule composition
3. Create a CIDR range function, it will return a range of `min` and `max` IP addresses, works well with rule composition
4. Add test and docs for these 3 features

**IP Converter Implementation**
IP addresses (string) is converted to their decimal format, using a simple calculation, it should be very efficient to run in production

Example composition `gt(ip("192.168.1.0"), ip("192.168.0.255"))` will return `1f`

**CIDR implementation**
Based on https://stackoverflow.com/questions/62124452/converting-cidr-notation-into-ip-range-without-other-library

The implementation returns a Range of Long, using a similar boilerplate as the range function `contains([min:max], value)`, so it can be strung together nicely in rule composition

Example composition `not(contains(cidr("254.200.222.210/23"), ip("254.200.223.255")))` will return `0f`

File: ffs-shared/env/src/commonMain/kotlin/env/EnvConst.kt [hunk 4]
```diff
@@ -8,3 +8,4 @@ const val ENV_USER_EMAIL = "user.email"
 const val ENV_DEVICE_NAME = "device.name"
 const val ENV_DEVICE_OS = "device.os"
 const val ENV_DEVICE_LOCALE = "device.locale"
+const val ENV_DEVICE_IP_ADDRESS = "device.ip_address"
```
Comment (goncalossilva) [inline ffs-shared/env/src/commonMain/kotlin/env/EnvConst.kt]:
```suggestion
const val ENV_DEVICE_IP_ADDRESS = "device.ip-address"
```

`internal.rollout-id` is the only other multi-word key, and it uses dashes.

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    var result = 0L
+    ipSegments.forEachIndexed { i, value ->
+        result += value * (256f.pow(3 - i)).toLong()
+    }
+
+    return result
+}
+
+fun ipv4ToNumeric(ipAddress: String): Long {
+    val splittedIpv4 = ipAddress.split('.').map { it.toShort() }
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Why `Short` and not `Byte`?

Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Why `Short` and not `Byte`? Octets are between 0 and 255.

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    if (splittedIpv4.size != 4 || splittedIpv4.any { it < 0 || it > 255 }) {
+        throw IllegalArgumentException("invalid IPv4 format")
+    }
+
+    return calculateIpv4Value(splittedIpv4)
+}
+
+fun cidrv4ToRange(cidr: String): LongRange {
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Could we inline this as a private function inside the `Cird` data class, since it's only used there? We use `calculateIpv4Value`, but we can replace it with: `octets.reduce { result, octet ->  (result shl 8) + octet }`

Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Same for `ipv4ToNumeric` inside `Ip`.

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    val mask = 0xFFFFFFFF shl (32 - subnetMask.toByte())
+    val subnet = arrayOf(
+        (mask ushr 24).toShort(),
+        (mask shr 16 and 0xff).toShort(),
+        (mask shr 8 and 0xff).toShort(),
+        (mask and 0xff).toShort()
+    )
+
+    val splittedIpv4 = baseIp.split('.')
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]:
```suggestion
    val octets = baseIp.split('.')
```

It's their name. :)

Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Let's do all validation at once, i.e., move this next to `if (!cidr.contains('/'))` above?

File: ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt [hunk 8]
```diff
@@ -507,5 +537,19 @@ private sealed class RuleExpr<out T> {
...
+        private fun delegateRangeToCollection(
+            range: LongRange
+        ): Collection<Long> = object : ClosedRange<Long> by range, Collection<Long> {
+            override val size: Int get() = (endInclusive - start).toInt()
+            override fun containsAll(elements: Collection<Long>) = elements.all {
+                range.contains(it)
+            }
+            override fun contains(value: Long): Boolean {
+                return range.contains(value)
+            }
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt]:
```suggestion
            override fun contains(value: Long) = range.contains(value)
```

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 9]
```diff
@@ -281,6 +294,58 @@ class RuleEvalTest {
...
 
+    @Test
+    fun ip() {
+        assertEquals(167772160f, eval("""ip("10.0.0.0")"""))
+        assertEquals(2560f, eval("""ip("0.0.10.0")"""))
+        assertEquals(2165248f, eval("""ip("0.33.10.0")"""))
+        assertEquals(3232238083f, eval("""ip("192.168.10.3")"""))
+        assertEquals(4294967041f, eval("""ip("255.255.255.1")"""))
+
+        // Test IP composition
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]:
```suggestion
```

Reading the code is enough.

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 10]
```diff
@@ -281,6 +294,58 @@ class RuleEvalTest {
...
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.0.0")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.-1")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.256")""") }
+        assertFailsWith<IllegalArgumentException> { eval("""ip("10.0.0.a")""") }
+    }
+
+    @Test
+    fun cidr() {
+        // testing reference from https://mxtoolbox.com/subnetcalculator.aspx
+
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]:
```suggestion
```

Subnets are just math, and many tools exist, future readers should trust this data!

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 11]
```diff
@@ -281,6 +294,58 @@ class RuleEvalTest {
...
+        // this syntax is not supported, will always return 0
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]:
```suggestion
```

The test already illustrates this. If this is a corner case, should we move it near the bottom, where the “exceptional” use cases are illustrated?

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt (hunk 10 above)
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]:
```suggestion
```

Subnets are just math, and many tools exist, future readers should trust this data!

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt [hunk 12]
```diff
@@ -281,6 +294,58 @@ class RuleEvalTest {
...
+        assertEquals(0f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.255.16"))"""))
+        assertEquals(0f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.12"))"""))
+
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.223.255"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("254.200.222.210/23"), ip("254.200.222.188"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.15"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.0"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.10/28"), ip("192.167.233.6"))"""))
+        assertEquals(1f, eval("""contains(cidr("192.167.233.11/32"), ip("192.167.233.11"))"""))
```
Comment (goncalossilva) [inline ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]: Let's have a couple of tests using cidr at the beginning of the range? E.g., does `0.0.0.0/0` work and include all IPv4 addresses?

**Review by goncalossilva — COMMENTED** [review]
_Excellent_ iteration, @kemasdimas! And great extension to ranges, it makes them a lot more useful. The design feels almost final. I've left a few more comments, but mostly small considerations.

Once we're ready, I'm thinking of squashing and merging, but if you prefer rebasing locally to tidy up the history, feel free to do so.

File: ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt (hunk 12 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt]: Yes, missed this one, will add the test 😊

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 7 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: I have a better proposal, what if we set `/32` as a fallback, so when a user doesn't put complete CIDR format, e.g `10.0.0.0` we will assume it to be `10.0.0.0/32`

Which essentially the same as checking a single IP address

What do you think?

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 6 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Sure we can do it inline inside Data Class, but we still need to have the `calculateIpv4Value` function, because we're using `List<Short>` and `reduce` only supports calculation in the same data type

Unless we're doing `.map` to `Long` beforehand, which will add more iteration to the calculation but will have a cleaner looking code

And considering we're using the exact same calculation 3 times, I think it's better to have a separate function for this

What do you think?

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 5 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: My consideration to use `Short` was to make the code look cleaner without needing to do additional calculations to handle the negative `Byte` value

But I've changed it to UByte and it seems to do the job 😊

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 6 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Calculation function will look like this
```
private fun calculateIpv4Value(octets: List<UByte>): Long {
    return octets.map { it.toLong() }.reduce { result, octet -> (result shl 8) + octet }
}
```

**Comment by kemasdimas** [comment]
Ooops, accidentally closed this pull request when cleaning up my repo fork. 😓

I will create a squashed commit that incorporates @goncalossilva recent comments 😊

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 6 above)
Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: > Unless we're doing .map to Long beforehand, which will add more iteration to the calculation but will have a cleaner looking code

Agree we shouldn't, it's wasteful. But I'm puzzled by `List<Short>`. Why can't we use `List<UByte>` everywhere?

> And considering we're using the exact same calculation 3 times, I think it's better to have a separate function for this

Definitely. I'm just struggling with its location!

The current structure leverages the hierarchy to place common functions in the top-most common ancestor's `Companion` object. That would be within `FunctionExpr.Companion`.

What I'm wondering is if we should have an `IpExpr` class with `Ip` and `Cidr` subclasses to approach it consistently, or if we could keep it simple and just have that method in FunctionExpr's Companion and use it. I tend to prefer the former, but would be OK with both.

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 7 above)
Comment (goncalossilva) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: Love it! 👌

File: ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt (hunk 6 above)
Comment (kemasdimas) [reply ffs-shared/rule/src/commonMain/kotlin/rule/IpParser.kt]: > Agree we shouldn't, it's wasteful. But I'm puzzled by List<Short>. Why can't we use List<UByte> everywhere?

Yes, we can use `UByte` instead of `Short`, and I've updated the latest commit to using `UByte` 😊 I was having an issue with value casting between unsigned to signed at first (getting failed tests results), so I resorted to `Short` 

> What I'm wondering is if we should have an IpExpr class with Ip and Cidr subclasses to approach it consistently, or if we could keep it simple and just have that method in FunctionExpr's Companion and use it. I tend to prefer the former, but would be OK with both.

`IpExpr` sounds good, it will contain all of the IPv4 and IPv6 functions within 1 place, so it will be easier to maintain
